### PR TITLE
feat: LT-1028 - Document Generation

### PR DESCRIPTION
**Pull request by BenAndrew-Kainos**

### CODE REVIEW

File: cases/forms/generate_document.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+from django.urls import reverse_lazy
+
+from lite_content.lite_internal_frontend import strings
+from lite_content.lite_internal_frontend.cases import GenerateDocumentsPage
+from lite_forms.components import Form, RadioButtonsImage, Option, BackLink
+
+
+def select_template_form(templates, case_id):
+    options = [
```
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply cases/forms/generate_document.py]: Sure :)

File: ui_automation_tests/pages/generate_document_page.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+from ui_automation_tests.shared.BasePage import BasePage
+
+
+class GeneratedDocument(BasePage):
+    preview = ".app-letter-preview__page"  # css
```
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply ui_automation_tests/pages/generate_document_page.py]: Done

File: ui_automation_tests/pages/application_page.py [hunk 3]
```diff
@@ -17,6 +17,7 @@ class ApplicationPage(BasePage):
     CASE_NOTE_DATE_TIME = ".app-activity__item .govuk-hint"  # css
     CASE_NOTE_CHARACTER_WARNING = "case_note-warning"  # id
     DOCUMENTS_BTN = '[href*="documents"]'  # css
+    GENERATE_DOCUMENTS_BTN = '[href*="generate-document"]'  # css
```
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply ui_automation_tests/pages/application_page.py]: Sure

File: templates/case/documents.html [hunk 4]
```diff
@@ -23,20 +26,24 @@ <h2 class="govuk-heading-l">{{ title }}</h2>
...
 						<th scope="row" class="govuk-table__header">{{ document.name }}</th>
+                        <td class="govuk-table__cell">{{ document.type.value }}</td>
 						<td class="govuk-table__cell">{{ document.description }}</td>
-						<td class="govuk-table__cell">{{ document.user.first_name }} {{ document.user.last_name }} at {{ document.created_at|str_date }}</td>
+						<td class="govuk-table__cell">{{ document.user.first_name }} {{ document.user.last_name }}</td>
+                        <td class="govuk-table__cell">{{ document.created_at|str_date }}</td>
 						<td class="govuk-table__cell">
 							{% if document.safe == True %}
-								<a class="govuk-link lite-link--icon govuk-link--no-visited-state" href="{% url 'cases:document' case.id document.metadata_id %}">
+								<a id="download" class="govuk-link lite-link--icon govuk-link--no-visited-state" href="{% url 'cases:document' case.id document.metadata_id %}">
```
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply templates/case/documents.html]: Done

File: templates/letter_templates/add_letter_paragraphs.html [hunk 5]
```diff
@@ -43,7 +43,7 @@ <h1 class="govuk-fieldset__heading">
             </button>
         </form>
     {% else %}
-        {% include "includes/information-text.html" with text='There aren\'t any letter paragraphs left to add.' %}
+        {% include "includes/information-text.html" with text="There aren't any letter paragraphs left to add." %}
```
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply templates/letter_templates/add_letter_paragraphs.html]: Done

File: templates/letter_templates/letter_templates.html [hunk 6]
```diff
@@ -47,7 +47,7 @@ <h2 class="govuk-heading-l">{% block title %}{% get_string 'letter_templates.let
 			</tbody>
 		</table>
 	{% else %}
-		{% include "includes/information-text.html" with text='There aren\'t any letter templates' %}
+		{% include "includes/information-text.html" with text="There aren't any letter templates" %}
```
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply templates/letter_templates/letter_templates.html]: Done

**Comment by efahy** [comment]
Minor comment: When I start typing a token in paragraph text and the matching entries are listed, it matches any entries that contain the letters I typed even if they  are not consecutive.  e.g. if I type "{raise" it gives me `{{open_application.application.organisation.vat_number}}` as a possible match.  Is there any way to only match entries that contain what I've typed exactly as I typed it?

**Comment by efahy** [comment]
Currently not possible to test paragraph token substitution because none of the tokens are up to date and won't get substituted according to Ben.  Further, due to the validation on paragraph creation, its only possible to add a token from the list and not possible to add a working one.  Please update list of tokens to include at least one working token for testing and demo purposes

**Comment by efahy** [comment]
Depending on the length of paragraph text, the table at the bottom of the ECJU Query letter template gets split across pages:
image

File: cases/urls.py [hunk 7]
```diff
@@ -73,4 +73,12 @@
     path("<uuid:pk>/review-goods/", goods.ReviewGoods.as_view(), name="review_goods"),
     # ex: /cases/<uuid:pk>/review-goods-clc/
     path("<uuid:pk>/review-goods-clc/", goods.ReviewGoodsClc.as_view(), name="review_goods_clc"),
+    # ex: /cases/<uuid:pk>/generate-document/
+    path("<uuid:pk>/generate-document/", generate_document.SelectTemplate.as_view(), name="generate_document"),
+    # ex: /cases/<uuid:pk>/generate-document/<uuid:tpk>/preview
```
Comment (markm5A8Z9F43913A) [inline cases/urls.py]: Examples be gone

File: core/builtins/custom_tags.py [hunk 8]
```diff
@@ -23,11 +23,33 @@ def get_const_string(value):
...
     warnings.warn("Reference constants from strings directly, only use LCS in HTML files", Warning)
     try:
-        return getattr(strings, value)
+        path = value.split(".")
+        # Get initial object from strings.py
+        path_object = getattr(strings, path[0])
+        return get(path_object, path[1:]) if len(path) > 1 else path_object
     except AttributeError:
-        return ""
+        return "STRING_NOT_FOUND"
```
Comment (markm5A8Z9F43913A) [inline core/builtins/custom_tags.py]: Content?

File: picklists/helpers.py [hunk 9]
```diff
@@ -1,11 +1,32 @@
-from django.template import Context, TemplateSyntaxError
+from django.template import Context, TemplateSyntaxError, Engine
 
 from letter_templates.context_variables import get_context_variables
-from letter_templates.helpers import template_engine_factory, InvalidVarException
+
+
+class InvalidVarException(Exception):
```
Comment (markm5A8Z9F43913A) [inline picklists/helpers.py]: As in the other pull request, Exceptions in a separate file would be great.

File: templates/case/documents.html [hunk 10]
```diff
@@ -23,20 +26,24 @@ <h2 class="govuk-heading-l">{{ title }}</h2>
 			<thead class="govuk-table__head">
 				<tr class="govuk-table__row">
 					<th scope="col" class="govuk-table__header">Name</th>
+                    <th scope="col" class="govuk-table__header">Type</th>
 					<th scope="col" class="govuk-table__header">Description</th>
-					<th scope="col" class="govuk-table__header">Uploaded by</th>
+					<th scope="col" class="govuk-table__header">Added by</th>
+                    <th scope="col" class="govuk-table__header">Date</th>
```
Comment (markm5A8Z9F43913A) [inline templates/case/documents.html]: Guess what I see here

Comment (markm5A8Z9F43913A) [reply templates/case/documents.html]: Go through the whole file please

File: templates/case/queries/clc-query-case.html [hunk 11]
```diff
@@ -21,22 +27,25 @@
...
-		<div class="lite-app-bar__controls">
-			{% if case.query.good.status.key != 'verified' and 'REVIEW_GOODS' in permissions %}
-				<a role="button" draggable="false" id="respond" class="govuk-button" href="{% url 'cases:respond_to_clc_query' case.id %}">
-					Respond to Query
-				</a>
-			{% endif %}
+			<div class="lite-app-bar__controls">
+                {% if case.query.good.status.key != 'verified' and 'REVIEW_GOODS' in permissions %}
+                    <a role="button" draggable="false" id="respond" class="govuk-button" href="{% url 'cases:respond_to_clc_query' case.id %}">
+                        Respond to Query
```
Comment (markm5A8Z9F43913A) [inline templates/case/queries/clc-query-case.html]: Content

File: ui_automation_tests/pages/generate_document_page.py [hunk 12]
```diff
@@ -0,0 +1,19 @@
+from ui_automation_tests.shared.BasePage import BasePage
+
+
+class GeneratedDocument(BasePage):
+    PREVIEW = ".app-letter-preview__page"  # css
```
Comment (markm5A8Z9F43913A) [inline ui_automation_tests/pages/generate_document_page.py]: Is there a reason we can't use an ID for this?

File: ui_automation_tests/pages/letter_templates.py [hunk 13]
```diff
@@ -9,7 +9,7 @@ class LetterTemplates(BasePage):
     ADD_LETTER_PARAGRAPH_BUTTON = '[value="add_letter_paragraph"]'  # CSS
     ADD_LETTER_PARAGRAPHS_BUTTON = '[value="return_to_preview"]'  # CSS
     PREVIEW_BUTTON = "button-preview"  # ID
-    PREVIEW_TEMPLATE_PARAGRAPHS = ".border-black.padding"  # CSS
+    PREVIEW = ".app-letter-preview__page"  # CSS
```
Comment (markm5A8Z9F43913A) [inline ui_automation_tests/pages/letter_templates.py]: Is there a reason we can't use an ID for this?

File: core/builtins/custom_tags.py [hunk 14]
```diff
@@ -23,11 +23,33 @@ def get_const_string(value):
...
+            # Search the object for the next property in `nested_properties_list`
+            return get(object, nested_properties_list[1:])
+
     warnings.warn("Reference constants from strings directly, only use LCS in HTML files", Warning)
     try:
-        return getattr(strings, value)
+        path = value.split(".")
+        # Get initial object from strings.py
+        path_object = getattr(strings, path[0])
+        return get(path_object, path[1:]) if len(path) > 1 else path_object
```
Comment (markm5A8Z9F43913A) [inline core/builtins/custom_tags.py]: This `try` block seems a bit too long. Which line can you get AttributeError on? Keep that in, and everything else should go before/after

File: templates/case/queries/end_user_advisory.html [hunk 15]
```diff
@@ -3,9 +3,15 @@
...
 {% block back_link %}
-	<a href="{% url 'cases:cases' %}?queue={{ queue.queue.id }}" class="govuk-back-link">
-		Back to {{ queue.queue.name }}
-	</a>
+	<div class="govuk-breadcrumbs">
+		<ol class="govuk-breadcrumbs__list">
+			<li class="govuk-breadcrumbs__list-item">
+				<a class="govuk-breadcrumbs__link" href="{% url 'cases:cases' %}?queue={{ queue.queue.id }}">{% if queue_name %}{{ queue_name }}{% else %}{% lcs 'QUEUE_ALL_CASES' %}{% endif %}</a>
+			</li>
+			<!-- TODO -->
```
Comment (markm5A8Z9F43913A) [inline templates/case/queries/end_user_advisory.html]: What's missing?

File: ui_automation_tests/step_defs/test_generated_documents.py [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+
+@then("I see the generated document preview")
+def generated_document_preview(driver, context):
+    generated_document_page = GeneratedDocument(driver)
+    assert generated_document_page.preview_is_shown()
+    paragraphs = generated_document_page.get_document_text()
+    for paragraph in context.document_template_paragraph_text:
+        assert paragraph in paragraphs
+    # When context variables are filled in at some future point also check that no {{ __ }} are present
```
Comment (markm5A8Z9F43913A) [inline ui_automation_tests/step_defs/test_generated_documents.py]: Is there a Jira ticket you can add this comment to instead?

File: templates/case/queries/clc-query-case.html [hunk 17]
```diff
@@ -3,9 +3,15 @@
...
 {% block back_link %}
-	<a href="{% url 'cases:cases' %}?queue={{ queue.queue.id }}" class="govuk-back-link">
-		Back to {{ queue_name }}
-	</a>
+	<div class="govuk-breadcrumbs">
+		<ol class="govuk-breadcrumbs__list">
+			<li class="govuk-breadcrumbs__list-item">
+				<a class="govuk-breadcrumbs__link" href="{% url 'cases:cases' %}?queue={{ queue.queue.id }}">{% if queue_name %}{{ queue_name }}{% else %}{% lcs 'QUEUE_ALL_CASES' %}{% endif %}</a>
+			</li>
+			<!-- TODO -->
```
Comment (markm5A8Z9F43913A) [inline templates/case/queries/clc-query-case.html]: What's missing?

File: core/builtins/custom_tags.py (hunk 8 above)
Comment (BenAndrew-Kainos) [reply core/builtins/custom_tags.py]: This is what's shown when lite-content isn't found, so we shouldn't move this to lite-content./ I'll make it a constant at the top instead

File: cases/urls.py (hunk 7 above)
Comment (BenAndrew-Kainos) [reply cases/urls.py]: Done

File: core/builtins/custom_tags.py (hunk 8 above)
Comment (BenAndrew-Kainos) [reply core/builtins/custom_tags.py]: Done this ^

File: picklists/helpers.py (hunk 9 above)
Comment (BenAndrew-Kainos) [reply picklists/helpers.py]: Done

File: templates/case/documents.html (hunk 10 above)
Comment (BenAndrew-Kainos) [reply templates/case/documents.html]: Done

File: templates/case/queries/clc-query-case.html (hunk 11 above)
Comment (BenAndrew-Kainos) [reply templates/case/queries/clc-query-case.html]: Done

File: templates/case/queries/end_user_advisory.html (hunk 15 above)
Comment (BenAndrew-Kainos) [reply templates/case/queries/end_user_advisory.html]: Done

File: ui_automation_tests/step_defs/test_generated_documents.py (hunk 16 above)
Comment (BenAndrew-Kainos) [reply ui_automation_tests/step_defs/test_generated_documents.py]: Done

File: ui_automation_tests/pages/generate_document_page.py (hunk 12 above)
Comment (BenAndrew-Kainos) [reply ui_automation_tests/pages/generate_document_page.py]: Done

File: ui_automation_tests/pages/letter_templates.py (hunk 13 above)
Comment (BenAndrew-Kainos) [reply ui_automation_tests/pages/letter_templates.py]: Done

File: core/builtins/custom_tags.py (hunk 14 above)
Comment (BenAndrew-Kainos) [reply core/builtins/custom_tags.py]: Done

File: templates/case/queries/clc-query-case.html (hunk 17 above)
Comment (BenAndrew-Kainos) [reply templates/case/queries/clc-query-case.html]: Nothing. Removed TODO

File: templates/case/documents.html [hunk 18]
```diff
@@ -3,7 +3,7 @@
 {% load svg %}
 
 {% block back_link %}
-	<a href="{% url 'cases:case' case.id %}" class="govuk-back-link">{% get_string 'cases.back_to_case' %}</a>
+	<a href="{% url 'cases:case' case.id %}" class="govuk-back-link">{% lcs 'CASES.CaseDocumentsPage.back_link' %}</a>
```
Comment (jfaracik) [inline templates/case/documents.html]: Shouldn't these be in all caps?

Comment (BenAndrew-Kainos) [reply templates/case/documents.html]: Done

**Pull request merged**